### DISCUSSION TITLE
Payment Plans - How to Implement

### DISCUSSION THREAD

**Discussion by gregberns** [vendurehq/vendure#676] [Q&A]
Hi, just found this project while researching headless EComm options to move away from our existing custom platform.

Background: I'm re-building an e-learning platform that sells courses that can be a bit expensive for students to pay for outright. So payment plans are optionally provided: 1 time payment, 6 monthly payments, 12 monthly payments, $1 down, etc. Currently the student selects the products they want to purchase, then when they view the cart (or at checkout) they can change the payment options, see updated pricing, then check out.

I'm trying to understand if this would be possible to implement. My guess is that maybe a plugin (or a couple) could be used, but I haven't gotten deep enough to understand what the details might look like.

Any help to determine the viability of this would be really appreciated.
Greg

**Comment by michaelbromley** [comment]
Hi Greg,

This would be possible to support provided you have a payment provider which supports such a payment plan. Vendure itself it very flexible when it comes to different payment flows - the framework supplies the hooks for you to create your own integration. If you already have a payment provider in mind, and can point me to any docs about payment plans, I could advise in a little more detail.

**Reply by gregberns** [reply]
Currently we're using Stripe to charge credit cards. In our site we'll calculate the pricing schedule and display the first payment, then charge the first payment, and then separately store the payment schedule and have a nightly job which charges customers for that day.

I'll give a little better example of the current workflow for more context: A customer adds two items to the cart, each is $500, total of $1000. Customer then gets to the cart/checkout and can then select to pay in: one payment, or 6 /12 monthly payments, and then optionally pay $1 as a first payment. Once the option is selected, the cart is re-calculated and they see the initial payment set to $1000/6 = $166.67. 

The nightly job is obviously out of scope, and we could probably keep storing the payment schedule out of scope too and just do it again elsewhere. I was thinking there would be a couple pieces:

* A GraphQL plugin that accepts the type of payment plan to use for the cart, and attaches the 'payment_plan_type' to the cart metadata
* A plugin that modifies the cart's price based on the payment plan selected
* A way to persist the plan selected to the database - new column(s) to store the type

Those are just some ideas, but I have no idea what that might look like in Vendure. Thanks!

**Reply by michaelbromley** [reply]
Alright, thanks for the clarification. So to take up the three points you mention:

> A GraphQL plugin that accepts the type of payment plan to use for the cart, and attaches the 'payment_plan_type' to the cart metadata

This is exactly how the Vendure [payment integrations](https://www.vendure.io/docs/developer-guide/payment-integrations/) work - there is a GraphQL mutation [`addPaymentToOrder`](https://www.vendure.io/docs/graphql-api/shop/input-types/#paymentinput) which takes a `metadata` argument. This is e.g. where you _could_ specify the payment plan type. Another way to do it (and maybe more suitable considering your requirements) would be to specify a [custom field](https://www.vendure.io/docs/developer-guide/customizing-models/) on the Order entity, which can store the plan type, and even things like records on payments being taken over time.

> A plugin that modifies the cart's price based on the payment plan selected

Yes, this would be possible in a couple of ways. One feature we have is the [OrderItemPriceCalculationStrategy](https://www.vendure.io/docs/typescript-api/orders/order-item-price-calculation-strategy/) which can dynamically set the price based on your own logic. In this case you'd need to set the payment plan on the OrderLine entity custom fields rather than Order. 

Another approach would be to add [Surcharges](https://www.vendure.io/docs/typescript-api/entities/surcharge/) to the order at some point in the order process, after the plan type has been selected.

> A way to persist the plan selected to the database - new column(s) to store the type

As mentioned, custom fields are good for this. If you want to store richer data you may also [define your own entities](https://www.vendure.io/docs/plugins/plugin-examples/defining-db-entity/) and set up a relation to the Order or Customer.

All in all, I'd say what you want to do is totally achievable in Vendure - it will just take a bit of experimentation to best model the workflow that you want.

**Reply by agustif** [reply]
Hi @gregberns would love to know what you end up doing as I'm also interested on using Stripe's Payment Plans.

We maybe could collaborate in a stripe payment plugin if you want to.

Have a nice day both of you =)
